### ISSUE TITLE
Don't trigger SLF001 on access to a class member from an instance of the class

### ISSUE THREAD

**Issue by cbornet** [astral-sh/ruff#17197]
### Summary

SLF001 triggers for this code:
```python
class A:
    def foo(self):
        do_something()

    def is_foo_overridden(self):
        """Returns true if the subclass of A overrides the foo method."""
        return type(self).foo is A.foo  # SLF001
```

It would be nice if it didn't trigger.

**Comment by trag1c** [comment]
This doesn't trigger SLF001 (the rule is about private member access, maybe you're thinking of something else?). Matter of fact there are zero violations on that line for latest ruff release with `--select ALL --preview` 😅

**Comment by ntBre** [comment]
This looks like it could be a further extension to the work in https://github.com/astral-sh/ruff/pull/16149.

**Comment by tpoliaw** [comment]
This appears to also cause problems when using Self or `"StringName"` to avoid dependency issues, eg

```python
from typing import Self

class Foo:
    def __init__(self, bar: int):
        self._bar = bar
    def stringed(self, other: "Foo") -> int:
        return self._bar + other._bar
    def selfed(self, other: Self) -> int:
        return self._bar + other._bar
```
Running `uvx ruff check --select SLF001 /tmp/demo.py` gives

```
SLF001 Private member accessed: `_bar`
 --> /tmp/demo.py:7:28
  |
5 |         self._bar = bar
6 |     def stringed(self, other: "Foo") -> int:
7 |         return self._bar + other._bar
  |                            ^^^^^^^^^^
8 |     def selfed(self, other: Self) -> int:
9 |         return self._bar + other._bar
  |

SLF001 Private member accessed: `_bar`
 --> /tmp/demo.py:9:28
  |
7 |         return self._bar + other._bar
8 |     def selfed(self, other: Self) -> int:
9 |         return self._bar + other._bar
  |                            ^^^^^^^^^^
  |

Found 2 errors.
```

where I wouldn't expect either to give errors
